esmvaltool/preprocessor/_mask.py

```diff
@@ -292,23 +297,31 @@ def mask_threshold(mycube, threshold):
 
 def mask_fillvalues(cubes, threshold_fraction, min_value=-1.e10,
                     time_window=1):
+    """Get the final fillvalues mask"""
     # function idea copied from preprocess.py
 
     # Ensure all cubes have masked arrays
     for cube in cubes:
         cube.data = np.ma.fix_invalid(cube.data, copy=False)
 
     # Get the fillvalue masks from all models
-    masks = (_get_fillvalues_mask(cube, threshold_fraction, min_value,
-                                  time_window) for cube in cubes)
+    masks = [_get_fillvalues_mask(cube, threshold_fraction, min_value,
+                                  time_window) for cube in cubes]
 
     # Combine all fillvalue masks
     combined_mask = None
     for mask in masks:
         if not np.all(mask):  # remove masks that mask everything
             if combined_mask is None:
                 combined_mask = np.zeros_like(mask)
-            combined_mask |= mask
+            if len(mask.shape) == 3:   # dig deeper in plevs
+                combined_copy = combined_mask.copy()
+                combined_mask |= mask
+                for i in range(mask.shape[0]):
+                    if np.all(combined_mask[i]):
+                        combined_mask[i] = combined_copy[i]
```

Comment:
Hi @bouweandela -- first, many thanks and beer your way for the very detailed review and comments! I have already committed the revised version on Friday evening but sorry for not having replied in line, I was keen on going to the pub :) So here's the comments -- for this one, the check is done to see if at pressure level "i" all values are masked (ie the vertical interpolation failed) and if so, just grab the previous valid mask and ignore this one. Ignoring the single mask does not preserve the previous combined mask that has to be propagated, if valid

Reply:
Just to stress the main point: this is absolutely needed because there may be other pressure levels "i+j" where the mask data is valid and the mask has to be added to the combined_mask hence the overarching statement combined_mask |= mask, but revert to the previous valid mask only the "i" plev where it's invalid -- this thing gave us a few days worth of headaches, man :)

Reply:
@valeriupredoi  Ok, I understand now why you skip masking if data for a pressure level is fully masked. However, I think it would be better to perform the check before applying the mask, so that you do not need to copy. A more efficient implementation would be:
```
    # Combine all fillvalue masks
    combined_mask = None
    for mask in masks:
        if combined_mask is None:
            combined_mask = np.zeros_like(mask)
        # Apply mask only if it doesn't mask everything
        n_dims = len(mask.shape)
        if n_dims in (1, 2) and ~np.all(mask):
            combined_mask |= mask
        elif n_dims == 3:
            # Select only valid pressure levels
            valid = ~np.all(mask, axis=(1, 2))
            combined_mask[valid] |= mask[valid]
        else:
            raise NotImplementedError("Unable to handle {} dimensional data"
                                      .format(n_dims))
```

Reply:
@bouweandela solution gleefully accepted and tested, works fine, cheers! no visible improvement in memory intake but it's nicer not to copy stuffs around :)